### PR TITLE
ARTEMIS-1478 - ActiveMQ Artemis to Kafka Bridge

### PR DESCRIPTION
**Pull request by michaelandrepearce**
Initial Kafka Bridge for Apache ActiveMQ Artemis to Apache Kafka, using Serivce Connector interface
Support to handle Core or AMQP based protocol over kafka.
Ensured for Core TextMessage can map to StringSerializer for a consumer
Ensure for Core ByteMessage can map to ByteArraySerializer for a consumer 
Kafka Serdes to support kafka consumers to consumer Core or AMQP back to CoreMessage, ProtonMessage, or JMSMessage
Added Documentation
Add integration tests

### CODE REVIEW

File: docs/user-manual/en/kafka-bridges.md [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+
+This maps the Message properties to Record headers.
+And then maps the payload binary as the Record value, encoding TextMessage
+
+This makes it easy to consume from Kafka using default deserializers
+TextMessage using StringDeserializer and
+ByteMessage using BytesDeserializer.
+
+Also to note, if you serialized an Apache Avro object into a byte array for the ByteMessage you could deserialize using an e
+
```
Comment (ppatierno) [inline docs/user-manual/en/kafka-bridges.md]: is this a broken line ?

Comment (michaelandrepearce) [reply docs/user-manual/en/kafka-bridges.md]: Good spot , Looks like something didn’t commit from my local properly I’ll correct.

Comment (michaelandrepearce) [reply docs/user-manual/en/kafka-bridges.md]: pushed, actually turns out i did just actually stop mid-sentence, during writing the doc, i have completed the section now.

File: integration/activemq-kafka/activemq-kafka-bridge/src/main/java/org/apache/activemq/artemis/integration/kafka/bridge/KafkaProducerBridge.java [hunk 2]
```diff
@@ -0,0 +1,484 @@
...
+         Integer partition = null;
+         SimpleString groupdID = message.getGroupID();
+         if (groupdID != null) {
+            List partitions = kafkaProducer.partitionsFor(topicName);
+            int numPartitions = partitions.size();
+            partition = Utils.toPositive(Utils.murmur2(groupdID.getData())) % numPartitions;
+         }
+
+         //Compaction/Record Key (equivalent to Artemis LVQ)
+         SimpleString key = message.getLastValueProperty();
```
Comment (ppatierno) [inline integration/activemq-kafka/activemq-kafka-bridge/src/main/java/org/apache/activemq/artemis/integration/kafka/bridge/KafkaProducerBridge.java]: I'm not an expert on Artemis but can you explain me why the LVQ is used as key for partitions ?

File: integration/activemq-kafka/activemq-kafka-bridge/src/main/java/org/apache/activemq/artemis/integration/kafka/bridge/KafkaProducerBridge.java [hunk 3]
```diff
@@ -0,0 +1,484 @@
...
+         synchronized (refs) {
+            refs.put(ref.getMessage().getMessageID(), ref);
+         }
+
+         Integer partition = null;
+         SimpleString groupdID = message.getGroupID();
+         if (groupdID != null) {
+            List partitions = kafkaProducer.partitionsFor(topicName);
+            int numPartitions = partitions.size();
+            partition = Utils.toPositive(Utils.murmur2(groupdID.getData())) % numPartitions;
```
Comment (ppatierno) [inline integration/activemq-kafka/activemq-kafka-bridge/src/main/java/org/apache/activemq/artemis/integration/kafka/bridge/KafkaProducerBridge.java]: It's the same way as the DefaultPartitioner works in Kafka, why do you re-used the logic here ?

File: integration/activemq-kafka/activemq-kafka-bridge/src/main/java/org/apache/activemq/artemis/integration/kafka/bridge/KafkaProducerBridge.java [hunk 4]
```diff
@@ -0,0 +1,484 @@
...
+         //Compaction/Record Key (equivalent to Artemis LVQ)
+         SimpleString key = message.getLastValueProperty();
+
+         ProducerRecord<String, Message> producerRecord =
+             new ProducerRecord<>(
+                   topicName,
+                   partition,
+                   message.getTimestamp(),
+                   key == null ? null : key.toString(),
+                   message);
```
Comment (ppatierno) [inline integration/activemq-kafka/activemq-kafka-bridge/src/main/java/org/apache/activemq/artemis/integration/kafka/bridge/KafkaProducerBridge.java]: Providing both partition and key (if it's not null), the Kafka producer will always use partition which takes precedence on the use of the key. It means that if groupId is specified, the LVQ (as key) won't be consider for addressing the partition. Is it exactly what you had in mind ?

File: integration/activemq-kafka/activemq-kafka-bridge/src/main/java/org/apache/activemq/artemis/integration/kafka/bridge/KafkaProducerBridge.java (hunk 2 above)
Comment (michaelandrepearce) [reply integration/activemq-kafka/activemq-kafka-bridge/src/main/java/org/apache/activemq/artemis/integration/kafka/bridge/KafkaProducerBridge.java]: Its not used for partitions, Key in Kafka is used for compaction. We use groupId (JMS MessageGroup) to sort partition see code above.

File: integration/activemq-kafka/activemq-kafka-bridge/src/main/java/org/apache/activemq/artemis/integration/kafka/bridge/KafkaProducerBridge.java (hunk 4 above)
Comment (michaelandrepearce) [reply integration/activemq-kafka/activemq-kafka-bridge/src/main/java/org/apache/activemq/artemis/integration/kafka/bridge/KafkaProducerBridge.java]: Yes key, is used for compaction in kafka primarily, we use groupId if present to primarily partition by to keep similar semantics as consumers in kafka consume by partition there for a group is honoured.

File: integration/activemq-kafka/activemq-kafka-bridge/src/main/java/org/apache/activemq/artemis/integration/kafka/bridge/KafkaProducerBridge.java (hunk 3 above)
Comment (michaelandrepearce) [reply integration/activemq-kafka/activemq-kafka-bridge/src/main/java/org/apache/activemq/artemis/integration/kafka/bridge/KafkaProducerBridge.java]: You don't have access to native headers in the partitioner. As such where serializers encode the payload into the value and map message headers (groupID) to kafka native headers, it is not available within kafka partitioner class.

**Comment by gemmellr** [comment]
Skimming the code quickly, I had to do a double take on a couple of occasions, and had some related observations.

There looked to be extensive use of implementation detail classes from the Qpid JMS client here. This seems inherently fragile as these classes are not intended for direct application usage in this fashion, and are entirely subject to change from release to release. Related, there also appeared to be some new class definitions within the org.apache.qpid namespace, which doesn't seem nice either.

Separately, I don't really understand why they would need/want to be used this way anyway. Presumably the Kafka message content in that case is an encoded AMQP message, so anything that can decode AMQP message would work, e.g the proton based bits also included. Its not clear to me why it woudl be desirable to involve JMS at all. Regardless, if you actually wanted a JMS message, why not consume it from the broker using the JMS client to begin with?

The logo feels like something only a standalone component would have.

**Comment by michaelandrepearce** [comment]
@gemmellr agreed on class def in names space this was because we couldn't create object because methods are package, as it notes it would be good if could get qpid to make them public to have them re-usable?

Yes anything can work, we are simply providing an implementation to make things easy out the box. 

Im happy to remove the QPID JMS one it is just for end users, giving as i said an out the box solution for them consumer side, but  its not actually needed by the bridge.

**Comment by tabish121** [comment]
Given that the code being used from Qpid JMS is internal implementation that has no promise of remaining stable and will not be made public I think that stuff should go.  

So far from what I see this doesn't seem like something that should go into the broker project but instead could be a stand alone project.   I'd prefer not to drag in Kafka dependencies into the distribution when not needed.

**Comment by michaelandrepearce** [comment]
@gemmellr I've removed the qpid JMS bit.

@tabish121 i see this no different to the vertx connecter service etc that was/is in 1.5.x range.

**Comment by gemmellr** [comment]
@michaelandrepearce Ah, I missed the note while skimming. The methods will presumably be at package visibility in keeping with the fact they are not intended to be used outside the client itself, so I don't see that changing. I think removing the JMS bits makes sense.

**Comment by michaelandrepearce** [comment]
@gemmellr have removed entirely to avoid issue.

**Comment by tabish121** [comment]
@michaelandrepearce apart from the fact that those are all removed now?

**Comment by michaelandrepearce** [comment]
there wasn't a JIRA to remove them, they just didn't get handled in the move to 2.x, and no ones done the work to update them to re-add them. Its not that we don't want connectors.

File: integration/activemq-kafka/activemq-kafka-bridge/src/main/java/org/apache/activemq/artemis/integration/kafka/bridge/KafkaProducerBridge.java (hunk 3 above)
Comment (ppatierno) [reply integration/activemq-kafka/activemq-kafka-bridge/src/main/java/org/apache/activemq/artemis/integration/kafka/bridge/KafkaProducerBridge.java]: "You don't have access to native headers in the partitioner" ...  I didn't get this point. Serialization and Partitioning are two different steps in the Kafka producer pipeline (first serialization, then partitioning). About the partitioner, if you pass the key to the producer send method, it applies exactly this : 

`Utils.toPositive(Utils.murmur2(keyBytes)) % numPartitions`

where `keyBytes` is just  `groupid`. Isn't it ?

Comment (michaelandrepearce) [reply integration/activemq-kafka/activemq-kafka-bridge/src/main/java/org/apache/activemq/artemis/integration/kafka/bridge/KafkaProducerBridge.java]: No keybytes is the LVQ property if set, as key is used for compacted topics.

Comment (michaelandrepearce) [reply integration/activemq-kafka/activemq-kafka-bridge/src/main/java/org/apache/activemq/artemis/integration/kafka/bridge/KafkaProducerBridge.java]: I have an idea of handling this with a custom partitioner and custom serialiser as an alternative

**Comment by clebertsuconic** [comment]
Vertx and aerogear were removed for lack of users also.  No one complained. If we had complainers we could put them back.

File: integration/activemq-kafka/activemq-kafka-bridge/src/main/java/org/apache/activemq/artemis/integration/kafka/bridge/KafkaProducerBridge.java (hunk 3 above)
Comment (michaelandrepearce) [reply integration/activemq-kafka/activemq-kafka-bridge/src/main/java/org/apache/activemq/artemis/integration/kafka/bridge/KafkaProducerBridge.java]: Updated so this logic is done with a partitioner (though it is more or less similar still) as we need to use groupID if present to partition by else we fall back to default of using key. But key must be LVQ for compaction topic reasons.

File: integration/activemq-kafka/activemq-kafka-protocols/activemq-kafka-amqp-protocol/src/main/java/org/apache/activemq/artemis/integration/kafka/protocol/amqp/AmqpMessageSerializer.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+public class AmqpMessageSerializer implements Serializer<Message> {
+
+   ProtonMessageSerializer protonMessageSerializer = new ProtonMessageSerializer();
+
+   @Override
+   public byte[] serialize(String topic, Message message) {
+      if (message == null) return null;
+      try {
+         AMQPMessage amqpMessage = CoreAmqpConverter.checkAMQP(message);
+         return protonMessageSerializer.serialize(topic, amqpMessage.getProtonMessage());
```
[the remark this answers is not in the document]
Comment (michaelandrepearce) [reply integration/activemq-kafka/activemq-kafka-protocols/activemq-kafka-amqp-protocol/src/main/java/org/apache/activemq/artemis/integration/kafka/protocol/amqp/AmqpMessageSerializer.java]: noted.

File: integration/activemq-kafka/activemq-kafka-bridge/src/main/java/org/apache/activemq/artemis/integration/kafka/bridge/GroupIdPartitioner.java [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+public class GroupIdPartitioner implements Partitioner {
+
+   DefaultPartitioner defaultPartitioner = new DefaultPartitioner();
+
+   @Override
+   public int partition(String topic, Object key, byte[] keyBytes, Object value, byte[] valueBytes, Cluster cluster) {
+      Message message = (Message) value;
+      final SimpleString groupdID = message.getGroupID();
+      //If GroupID (a.k.a JMSXMessageGroup) then to honour this we partition based on it.
```
[the remark this answers is not in the document]
Comment (michaelandrepearce) [reply integration/activemq-kafka/activemq-kafka-bridge/src/main/java/org/apache/activemq/artemis/integration/kafka/bridge/GroupIdPartitioner.java]: good point.

File: integration/activemq-kafka/activemq-kafka-bridge/src/main/java/org/apache/activemq/artemis/integration/kafka/bridge/ActiveMQKafkaLogger.java [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+ * each message id must be 6 digits long starting with 36, the 3rd digit donates the level so
+ *
+ * INF0  1
+ * WARN  2
+ * DEBUG 3
+ * ERROR 4
+ * TRACE 5
+ * FATAL 6
+ *
+ * so an INFO message would be 191000 to 191999
```
[the remark this answers is not in the document]
Comment (michaelandrepearce) [reply integration/activemq-kafka/activemq-kafka-bridge/src/main/java/org/apache/activemq/artemis/integration/kafka/bridge/ActiveMQKafkaLogger.java]: agreed.

File: docs/user-manual/en/kafka-bridges.md [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+Apache Kafka and get a ProtonMessage.
+
+`org.apache.activemq.artemis.integration.kafka.protocol.amqp.proton.ProtonMessageDeserializer`
+
+You can get these Apache Kafka, Serializers/Deserializers via maven using the following GAV coordinates:
+
+    <depedency>
+       <groupId>org.apache.activemq</groupId>
+       <artifactId>artemis-kafka-amqp-protocol</artifactId>
+       <version>2.4.0-SNAPSHOT</version>
```
[the remark this answers is not in the document]
Comment (michaelandrepearce) [reply docs/user-manual/en/kafka-bridges.md]: agreed.

File: docs/user-manual/en/kafka-bridges.md [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+This encodes the whole message into amqp binary protocol into the Record value.
+
+This is useful if you have something like
+
+   https://github.com/EnMasseProject/amqp-kafka-bridge
+
+As it allows messages to pump into kafka in amqp binary, to then consumed from kafka by clients using amqp binary protocol.
+
+Also provided are Apache Qpid Proton based Apache Kafka deserializers allowing you to consume from 
+Apache Kafka and get a ProtonMessage.
```
[the remark this answers is not in the document]
Comment (michaelandrepearce) [reply docs/user-manual/en/kafka-bridges.md]: will do.

File: docs/user-manual/en/kafka-bridges.md [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+    </depedency>
+
+
+#### AMQPMessageSerializer
+Can be set by using:
+    
+    <param key="value.serializer" value="org.apache.activemq.artemis.integration.kafka.protocol.amqp.AMQPMessageSerializer" />
+
+
+This encodes the whole message into amqp binary protocol into the Record value.
```
[the remark this answers is not in the document]
Comment (michaelandrepearce) [reply docs/user-manual/en/kafka-bridges.md]: will capitalise

File: docs/user-manual/en/kafka-bridges.md [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+
+`org.apache.activemq.artemis.integration.kafka.protocol.core.jms.CoreJmsMessageDeserializer`
+`org.apache.activemq.artemis.integration.kafka.protocol.core.CoreMessageDeserializer`
+
+You can get these Apache Kafka, Serializers/Deserializers via maven using the following GAV coordinates:
+
+    <depedency>
+       <groupId>org.apache.activemq</groupId>
+       <artifactId>artemis-kafka-core-protocol</artifactId>
+       <version>2.4.0-SNAPSHOT</version>
```
[the remark this answers is not in the document]
Comment (michaelandrepearce) [reply docs/user-manual/en/kafka-bridges.md]: agreed, will replace with 2.x.x

File: docs/user-manual/en/kafka-bridges.md [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+Taking and giving the best features of both broker technologies when and where needed for a flow of data.
+
+![ActiveMQ Artemis Kafka Bridge Clients](images/activemq-kafka-bridge-clients.png)
+
+The intent is this will be a two way bridge, but currently the flow is a single direction 
+from Apache ActiveMQ Artemis to Apache Kafka
+
+
+The source and target servers are remote making bridging suitable
+for reliably sending messages from one artemis cluster to kafka,
```
[the remark this answers is not in the document]
Comment (michaelandrepearce) [reply docs/user-manual/en/kafka-bridges.md]: will Capatilise

File: docs/user-manual/en/kafka-bridges.md [hunk 13]
```diff
@@ -0,0 +1,180 @@
+# Apache ActiveMQ Kafka Bridge
+
+The function of a bridge is to consume messages from a source queue in Apache ActiveMQ Artemis, 
+and forward them to a target topic, on a remote Apache Kafka server.
+
+By pairing Apache ActiveMQ Artemis and Apache Kafka with the bridge you could have a hybrid broker setup, 
+having a data flow with CORE, AMQP, MQTT clients, as well as now Kafka clients also. 
+Taking and giving the best features of both broker technologies when and where needed for a flow of data.
+
+![ActiveMQ Artemis Kafka Bridge Clients](images/activemq-kafka-bridge-clients.png)
```
[the remark this answers is not in the document]
Comment (michaelandrepearce) [reply docs/user-manual/en/kafka-bridges.md]: will change image

File: docs/user-manual/en/kafka-bridges.md [hunk 14]
```diff
@@ -0,0 +1,180 @@
...
+The source and target servers are remote making bridging suitable
+for reliably sending messages from one artemis cluster to kafka, 
+for instance across a WAN, to the cloud, or internet and where the connection may be unreliable.
+
+The bridge has built in resilience to failure so if the target server
+connection is lost, e.g. due to network failure, the bridge will retry
+connecting to the target until it comes back online. When it comes back
+online it will resume operation as normal.
+
+![ActiveMQ Artemis Kafka Bridge Diagram](images/activemq-kafka-bridge-diagram.png)
```
[the remark this answers is not in the document]
Comment (michaelandrepearce) [reply docs/user-manual/en/kafka-bridges.md]: will update image

**Comment by michaelandrepearce** [comment]
@gemmellr Thanks for the comments on the docs, code and images, have addressed these

File: docs/user-manual/en/kafka-bridges.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (michaelandrepearce) [reply docs/user-manual/en/kafka-bridges.md]: have updated image.

File: docs/user-manual/en/kafka-bridges.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (michaelandrepearce) [reply docs/user-manual/en/kafka-bridges.md]: have updated image.

**Comment by michaelandrepearce** [comment]
so if looked to take this route how easy is it to get a sub project of activemq artemis (e.g. a bit like NMS)  to hold the eco system tools/extensions such as this? e.g named project: ActiveMQ Artemis Service Connectors to hold these.

**Comment by michaelandrepearce** [comment]
For the interim until we sort seperate sub project so this isn’t lost I’d like to merge this still then split it out later.

**Comment by gemmellr** [comment]
I don't think that makes sense personally. If the end view is that it should be maintained outwith the broker repo/distribution, that is where it should go from the outset. In the interim folks can build it themselves.

In terms of establishing a location outside the broker repo, or getting more views on whether that is what should happen, perhaps you should start a discussion on the dev list about where it goes.

**Comment by cshannon** [comment]
I agree with @mtaylor that I think this should be maintained outside the broker for the reasons he already mentioned.

**Comment by michaelandrepearce** [comment]
Thanks guys for the time, I have kicked off back the dev mail list to discuss the idea of having a sub project for service connectors and other integrations like the spring one. Let’s discuss there.

**Comment by michaelandrepearce** [comment]
Closing this for now

**Pull request closed without merging**